src/resources/ble_esp32.cc

```diff
@@ -575,12 +578,24 @@ class ServiceContainer : public BleErrorCapableResource {
   ServiceContainer(BleResourceGroup* group, Kind kind)
       : BleErrorCapableResource(group, kind) {}
 
+  ~ServiceContainer() override {
+    while (!services_.is_empty()) {
+      auto service = services_.first();
+      group()->unregister_resource(service);
```

Comment:
Maybe add a comment that explains that unregistering will remove the element from linked list?

Reply:
done.